### ISSUE TITLE
tablets: implement recovery

### ISSUE THREAD

**Issue by piodul** [scylladb/scylladb#17705]
In case when a quorum of nodes is lost permanently or upgrade to raft failed, we have a [recovery procedure for group 0](https://opensource.docs.scylladb.com/stable/architecture/raft.html#raft-manual-recovery-procedure). This procedure has recently been extended to support topology on raft (in [this PR](https://github.com/scylladb/scylladb/pull/17077), the documentation is cooking [here](https://github.com/scylladb/scylladb/pull/17624)).

However, this extended procedure does not do anything special with regards to tablets and recovery on a cluster with tablets is not currently tested. The behavior of the procedure needs to be investigated and possibly adjusted to work with tablets.

**Comment by kbr-scylla** [comment]
@bhalevy @kostja @tgrabiec we need to decide what happens with tablets after quorum loss and recovery

**Comment by kbr-scylla** [comment]
cc @denesb

**Comment by kbr-scylla** [comment]
cc @mykaul

**Comment by mykaul** [comment]
Let's begin with understanding where we are at. What are the gaps.

**Comment by kbr-scylla** [comment]
@tgrabiec I wonder if it's possible to preserve availability of tablet-based tables during recovery. We can keep the contents of `tablets` metadata table intact during recovery, and load it into memory. We wouldn't do load balancing obviously because there's no topology coordinator. So at least when we immediately restart in recovery, it may be possible to do data path ops on tablet-based tables.

The question is what happens next. During recovery we revert to gossip-based operations and `removenode` the dead nodes. One potential problem is, of course, that tablet metadata won't be updated, we won't be able to move tablets during that operation. Not sure if that is a problem though.

Then we restart from recovery and perform upgrade-to-raft procedure again. This procedure would have to notice that contents of tablets metadata table is nonempty, and potentially inconsistent across nodes. It would have to somehow repair it and then remove the tablet replicas which are assigned to now-removed nodes.

**Comment by kbr-scylla** [comment]
See also https://docs.google.com/document/d/1KHiPRNwxpuVW46ir9OFfLDj2at68K_BcZH2a__0wkGs/edit#heading=h.hftkdz48zxm9

**Comment by bhalevy** [comment]
The headline of this issue is a bit terse.
Please describe what kind of recovery this issue is about.

**Comment by kbr-scylla** [comment]
Everything's in the first post --- we have the "manual Raft recovery procedure" (https://opensource.docs.scylladb.com/stable/architecture/raft.html#raft-manual-recovery-procedure) which has been updated to support Raft-based topology (https://github.com/scylladb/scylladb/pull/17624).

The issue is about extending the procedure with tablets support.

**Comment by kbr-scylla** [comment]
updated title

**Comment by bhalevy** [comment]
I doubt this can be done in the 6.0 timeframe.
@mykaul / @dani-tweig please consider postponing to 6.1

**Comment by kbr-scylla** [comment]
@bhalevy in that case it means **we do not have a way out from group 0 majority loss if we use tablets** other than restoring from backup :(

Documentation will have to be updated.

**Comment by mykaul** [comment]
That's quite an issue. I assume we won't have many tablets only setups, but that certainly is a candidate, once fixed, to a 6.0.x

**Comment by dani-tweig** [comment]
moved it to 6.0.1 for now.

**Comment by kbr-scylla** [comment]
We cannot leave documentation as it is for 6.0

**Comment by kbr-scylla** [comment]
>  I assume we won't have many tablets only setups,

It's not just "tablets only". It's sufficient to have just one tablet-based table and 99 vnode-based tables and this invalidates the RECOVERY procedure. We don't even know (nobody tested) if the existing RECOVERY procedure will work and restore the remaining 99 vnode-based tables, and what will happen to that one tablet-based table, will Scylla crash? Will it refuse queries to this table? Or enter some undefined behavior?

**Comment by mykaul** [comment]
> > I assume we won't have many tablets only setups,
> 
> It's not just "tablets only". It's sufficient to have just one tablet-based table and 99 vnode-based tables and this invalidates the RECOVERY procedure. We don't even know (nobody tested) if the existing RECOVERY procedure will work and restore the remaining 99 vnode-based tables, and what will happen to that one tablet-based table, will Scylla crash? Will it refuse queries to this table? Or enter some undefined behavior?

We'll need to figure it and fix then, sounds like a blocker to me.

**Comment by mykaul** [comment]
> > I assume we won't have many tablets only setups,
> 
> It's not just "tablets only". It's sufficient to have just one tablet-based table and 99 vnode-based tables and this invalidates the RECOVERY procedure. We don't even know (nobody tested) if the existing RECOVERY procedure will work and restore the remaining 99 vnode-based tables, and what will happen to that one tablet-based table, will Scylla crash? Will it refuse queries to this table? Or enter some undefined behavior?

Related: don't we have tests for recovery?

**Comment by kbr-scylla** [comment]
> Related: don't we have tests for recovery?

We do. They don't run in --tablets mode (obviously, because nobody extended RECOVERY to procedure to support tablets yet.)

**Comment by tgrabiec** [comment]
@kbr-scylla Can't quorum loss recovery work by keeping the cluster in raft mode, but reduce the members to the alive nodes, ensure that all are caught up before group0 is available, and then mark the dead nodes as "permanently dead" so that topology ops can proceed. Why do we fall back to gossip?

**Comment by kbr-scylla** [comment]
> but reduce the members to the alive nodes

Reduce members == Raft configuration change == must commit an entry to the log == needs quorum.

**Comment by bhalevy** [comment]
> > but reduce the members to the alive nodes
> 
> Reduce members == Raft configuration change == must commit an entry to the log == needs quorum.

Can't we edit the logs locally on each node, if they are all put into maintenance mode?

**Comment by kbr-scylla** [comment]
>  Can't we edit the logs locally on each node, if they are all put into maintenance mode?

The logs are sequences of commands and commands are sequences of bytes.

So first you would need a tool to deserialize the commands. It would have to understand everything that Scylla understands about these commands.

Even if you did manage to deserialize the commands, it is **very non-obvious** what to do next with these logs. The logs are part of a larger picture. Let's say you pick one of the logs and overwrite all others with it. But it could turn out that you picked wrong, because some other nodes already applied commands from another log that you didn't pick and the log you did pick does not have these commands. Or suppose you took the longest log. Again could be wrong, because a shorter log could've contained applied commands, while the longest log did not. You would have to analyze the combination of Raft terms and indexes of commands. Probably doable but not simple. Raft is not a simple algorithm.

That said I do want to create an external tool for recovery in the future so we can get rid of gossip-based recovery mode. The tool would probably not analyze Raft log, but look at the current states of remaining nodes (system.topology etc.) and reconcile them and determine some new initial state (with the dead nodes removed) and then re-initialize Raft on all nodes from scratch (empty log) but using this externally calculated state as the initial state. I have an issue for it: https://github.com/scylladb/scylladb/issues/18791 but with 6.1 milestone. For 6.0 it was simpler for us to extend the already-existing gossip-based RECOVERY mode (which we did way back for schema-on-Raft in 5.2). The tool needs to be designed, the design reviewed, and then the implementation. It will be a long process, I estimate it as `L` (a month) and put in Q3 plan for my team. Also the tool will still not automagically solve tablets -- the reconciliation process will have to calculate the tablet mapping and this is specific to how tablets work.

**Comment by bhalevy** [comment]
> > Can't we edit the logs locally on each node, if they are all put into maintenance mode?
> 
> The logs are sequences of commands and commands are sequences of bytes.
> 
> So first you would need a tool to deserialize the commands. It would have to understand everything that Scylla understands about these commands.
> 
> Even if you did manage to deserialize the commands, it is **very non-obvious** what to do next with these logs. The logs are part of a larger picture. Let's say you pick one of the logs and overwrite all others with it. But it could turn out that you picked wrong, because some other nodes already applied commands from another log that you didn't pick and the log you did pick does not have these commands. Or suppose you took the longest log. Again could be wrong, because a shorter log could've contained applied commands, while the longest log did not. You would have to analyze the combination of Raft terms and indexes of commands. Probably doable but not simple. Raft is not a simple algorithm.
> 
> That said I do want to create an external tool for recovery in the future so we can get rid of gossip-based recovery mode. The tool would probably not analyze Raft log, but look at the current states of remaining nodes (system.topology etc.) and reconcile them and determine some new initial state (with the dead nodes removed) and then re-initialize Raft on all nodes from scratch (empty log) but using this externally calculated state as the initial state. I have an issue for it: #18791 but with 6.1 milestone. For 6.0 it was simpler for us to extend the already-existing gossip-based RECOVERY mode (which we did way back for schema-on-Raft in 5.2). The tool needs to be designed, the design reviewed, and then the implementation. It will be a long process, I estimate it as `L` (a month) and put in Q3 plan for my team. Also the tool will still not automagically solve tablets -- the reconciliation process will have to calculate the tablet mapping and this is specific to how tablets work.

Waving my hands, I can see a tool that works at the cql level, reading the raft state from all nodes, deciding what to do, and applying that back onto all nodes using cql, all that with the nodes in maintenance mode so they can serve cql to their local tables.

**Comment by kbr-scylla** [comment]
Yes, that's how I imagine it as well. The largest design problem here is to how construct the new state based on the states collected from all live nodes, which might come from different points in time, e.g. if majority loss happened in the middle of topology change, or we did some schema changes while some node was partitioned, etc.

**Comment by tgrabiec** [comment]
On Fri, May 24, 2024 at 5:02 PM Kamil Braun ***@***.***>
wrote:

> Yes, that's how I imagine it as well. The largest design problem here is
> to how construct the new state based on the states collected from all live
> nodes, which might come from different points in time, e.g. if majority
> loss happened in the middle of topology change, or we did some schema
> changes while some node was partitioned, etc.
>

Wouldn't the leader election in the new quorum pick the node which has the
largest amount of commits? In that case, the new leader should have the
most up to date state (schema, topology), and other nodes will catch up
with it.

**Comment by kbr-scylla** [comment]
> We can shadow raft membership with a blacklist which comes from yaml for
recovery purposes, which will allow us to regain quorum. Consistency is
preserved assuming the nodes in the blacklist are never coming up.


Do you have a proof of correctness (safety and liveness) with Raft modified in this way?

I'm guessing (but didn't verify) that it could happen that no node in the present set of members is able to win leader election.

> Wouldn't the leader election in the new quorum pick the node which has the
largest amount of commits?

No.

**Comment by kbr-scylla** [comment]
>  Assuming all transitions are atomic (with the help of the commitlog), wouldn't reconciling the latest state on all nodes bring the cluster back into sanity?

I don't know.

**Comment by kbr-scylla** [comment]
@dani-tweig @mykaul @avikivity if this is not done to 6.0, **then we need to update the 6.0 documentation that in case of majority loss, restore-from-backup is the only remaining option**. It should be a release blocker -- either making RECOVERY mode work, or updating the documentation. We shouldn't release without doing either.

**Comment by dani-tweig** [comment]
> **in case of majority loss, restore-from-backup is the only remaining option**.

@tzach , @annastuchlik , We will need to add a limitation to 6.0 documentation.
Maybe another task in https://github.com/scylladb/scylladb/issues/18597 ?

**Comment by dani-tweig** [comment]
> Has the decision been made that RECOVERY is not going to be fixed in 6.0?

Yes

**Comment by mykaul** [comment]
> > Has the decision been made that RECOVERY is not going to be fixed in 6.0?
> 
> Yes

More accurately, not in 6.0.0, but I have high hopes it'll be fixed soon after.

**Comment by tzach** [comment]
With Tablets as default and no Raft recovery procedure for Tablets, I see little value for the recovery procedure for 6.0.

**Comment by tzach** [comment]
Till this is fixed, we need to rewrite the recovery procedure to create a new cluster and restore data from the backup.

**Comment by tgrabiec** [comment]
Here's a proposal for a solution, which requires no code changes for tablets: https://docs.google.com/document/d/1fgQKxJerzIpK-Ow07W8v-LKvm-KK1rFSF457Hw-dQsQ/edit

But it relies on general group0 recovery: https://docs.google.com/document/d/1Y7UZ2VOcQ5Un55wJvTunsaafihndAFP9mtOzTlfIOMI/edit

**Comment by kbr-scylla** [comment]
Blocked by https://github.com/scylladb/scylladb/issues/20657

**Comment by bhalevy** [comment]
@xemul should this issue still be assigned to you or be moved to @tgrabiec?

**Comment by xemul** [comment]
@bhalevy , it's "planning question", so Cc @regevran

**Comment by regevran** [comment]
If @tgrabiec can take it it would be better.

**Comment by avikivity** [comment]
Seems like @tgrabiec already proposed something in https://github.com/scylladb/scylladb/issues/17705#issuecomment-2333486698.

@tgrabiec please present it at the next raft meeting, or core-arch meeting, whichever you think is a better fit.

**Comment by tgrabiec** [comment]
> Seems like [@tgrabiec](https://github.com/tgrabiec) already proposed something in [#17705 (comment)](https://github.com/scylladb/scylladb/issues/17705#issuecomment-2333486698).
> 
> [@tgrabiec](https://github.com/tgrabiec) please present it at the next raft meeting, or core-arch meeting, whichever you think is a better fit.

Today's raft meeting is too soon, so let's make it core-arch. \cc @dani-tweig

**Comment by dani-tweig** [comment]
> Today's raft meeting is too soon, so let's make it core-arch. \cc [@dani-tweig](https://github.com/dani-tweig)

Next available arch meeting is on October 30th. Is that ok?

**Comment by tgrabiec** [comment]
Summary of the meeting:

- slides: https://docs.google.com/presentation/d/1E8EC-C92oLBbXfEY6b_-br47lvwDjZ9CjC2or_k0SNQ/edit#slide=id.g310296f5919_0_939
- Raft team is working on a tool to restore group0 availability and the most-recent group0 state from the surviving nodes
- Once we have that, there is nothing we have to do for tablet recovery, if we commit to preserving the "group0 causality property"
- The property is about ensuring that group0 state is committed locally before any effects of that state are mode locally
- schema changes preserve it, we issue read_barrier() before processing request which uses unknown schema
- tablet migration preserves it, topology coordinator executes token metadata barrier before executing actions associated with the stage.   
- There was no disagreement to proceed this way
- Kamil pointed out that this property is also necessary for correctness of other things, not just recovery
- Glab pointed out that if we rely on this property, we have to commit to having group0 replica on every node

**Comment by kbr-scylla** [comment]
> https://github.com/scylladb/scylladb/issues/17705#issuecomment-2450905078

cc @patjed41

**Comment by dani-tweig** [comment]
@kbr-scylla , is this item still blocked? (How so?) , or can the label be removed?
What are the next steps?

**Comment by kbr-scylla** [comment]
https://github.com/scylladb/scylladb/issues/17705#issuecomment-2355797632

Doing https://github.com/scylladb/scylladb/issues/20657 will automatically fix this item.

**Comment by mykaul** [comment]
@emdroid , @patjed41 - what's the status of this one?

**Comment by emdroid** [comment]
> [#17705 (comment)](https://github.com/scylladb/scylladb/issues/17705#issuecomment-2355797632)
> 
> Doing [#20657](https://github.com/scylladb/scylladb/issues/20657) will automatically fix this item.

@patjed41 as per the above comment and the #20657 being already done, I suppose we could close this one?

**Comment by patjed41** [comment]
Yes, the new recovery procedure works with tablets (and everything else built on the Raft-based topology).
